Exception Handling (#6)
Ada-Developers-Academy/core-unit-1: exception-handling/practice-exception-handling.md

# Practice: Exception Handling

## Directions

Complete all questions below.

## Practice

<!-- Question 1 -->

<!-- prettier-ignore-start -->
### !challenge

* type: code-snippet
* language: python3.6
* id: XwJz6l
* title: Exception Handling

### !question

> Given these tests, implement the function that will pass. Assume that the only invalid, claimed code school name is `"Ada Developers Academy"`.

Reviewer: Possible rephrase since we often say that we write code to make tests pass.
```suggestion
Given these tests, provide the function body that will make them pass. Assume that the only invalid, claimed code school name is `"Ada Developers Academy"`.
```

Author: heads up, i just wrote like a dozen questions that mirror this wording LOL

    result = claim_unreserved_code_school_name("Some Code School")
    assert result is True

def test_invalid_name():
    with pytest.raises(ValueError):
        claim_unreserved_code_school_name("Ada Developers Academy")
```

### !end-question
### !placeholder

```python
def claim_unreserved_code_school_name(name):
    pass
```
### !end-placeholder
### !tests
```python
import unittest
from main import *

class TestChallenge(unittest.TestCase):
    def test_valid_name(self):
        result = claim_unreserved_code_school_name("Some Code School")
        self.assertTrue(result)

    def test_invalid_name(self):
        with self.assertRaises(ValueError):
            claim_unreserved_code_school_name("Ada Developers Academy")
```
### !end-tests
### !explanation

An example of a working implementation:

```python
def claim_unreserved_code_school_name(name):
    if name == "Ada Developers Academy":
        raise ValueError
    return True
```

### !end-explanation
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question 2 -->

<!-- prettier-ignore-start -->
### !challenge

* type: code-snippet
* language: python3.6
* id: SkMsOl
* title: Exception Handling

### !question

> Given these tests, implement the function that will pass. Assume that any phone number that is a string is valid.

Reviewer: Similar rephrase as above.
```suggestion
Given these tests, provide the function body that will make them pass. Assume that any phone number that is a string is valid.
```

Hint: The function `type(phone_num)` will return `str` if `phone_num` is a string.

```python
def test_valid_phone_nums_return_true():
    phone_num = "777-888-9999"
    is_valid = is_phone_num_valid(phone_num)
    assert is_valid

def test_invalid_phone_nums_raise_error():
    phone_num = 777-888-9999
    with pytest.raises(ValueError):
        is_valid = is_phone_num_valid(phone_num)
```

### !end-question
### !placeholder

```python

> def is_phone_num_valid(phone_num):
>     return True

Reviewer: Use `pass` as in the previous example.
```suggestion
def is_phone_num_valid(phone_num):
    pass
```

```
### !end-placeholder
### !tests
```python
import unittest
from main import *

class TestChallenge(unittest.TestCase):
    def test_valid_phone_nums_return_true(self):
        phone_num = "777-888-9999"
        is_valid = is_phone_num_valid(phone_num)
        self.assertTrue(is_valid)

    def test_invalid_phone_nums_raise_error(self):
        phone_num = 777-888-9999
        with self.assertRaises(ValueError):
            is_phone_num_valid(phone_num)
```
### !end-tests
### !explanation

An example of a working implementation:

```python
def is_phone_num_valid(phone_num):
    if type(phone_num) is not str:
        raise ValueError
    return True
```

### !end-explanation
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question 3 -->

<!-- prettier-ignore-start -->
### !challenge

* type: code-snippet
* language: python3.6
* id: VLwvhu
* title: Exception Handling

### !question

> Given these tests, refactor this function so it will pass. Use `try ... except` syntax.

Reviewer: Similar rephrasing as before.
```suggestion
Given these tests, refactor this function to make them pass. Use `try ... except` syntax.
```

```python
def split_cost_evenly(total_cost, num_of_people, tip_percentage):
    cost_with_tip = total_cost * (1 + tip_percentage)
    cost_per_person = cost_with_tip / num_of_people
    return cost_per_person
```
### !end-placeholder
### !tests
```python
import unittest
from main import *

class TestChallenge(unittest.TestCase):
    def test_split_cost_evenly(self):
        cost = split_cost_evenly(100, 4, .5)
        self.assertAlmostEqual(cost, 37.5)

    def test_caught_zerodivisionerror_returns_zero(self):
        cost = split_cost_evenly(100, 0, .5)
        self.assertEqual(cost, 0)
```
### !end-tests
### !explanation

An example of a working implementation:

```python
def split_cost_evenly(total_cost, num_of_people, tip_percentage):
    try:
        cost_with_tip = total_cost * (1 + tip_percentage)
        cost_per_person = cost_with_tip / num_of_people
        return cost_per_person
    except ZeroDivisionError:
        return 0
```

### !end-explanation
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question 4 -->

<!-- prettier-ignore-start -->
### !challenge

* type: code-snippet
* language: python3.6
* id: j9kUX9
* title: Exception Handling

### !question

> Given these tests, refactor **_only_** the **`get_celebration_treat`** function so it will pass. Use `try ... except` syntax.

Reviewer: Similar rephrasing to above.
```suggestion
Given these tests, refactor **_only_** the **`get_celebration_treat`** function to make them pass. Use `try ... except` syntax.
```

In `get_celebration_treat`...

- If no error is raised, then `get_celebration_treat` returns `"celebration {thirteenth_donut}!"`, where `thirteenth_donut` is the result of `grab_thirteenth_donut()`
- If an `IndexError` is raised because of accessing something missing in `donuts`, then `get_celebration_treat` returns `"no donut :("`

```python
def test_get_celebration_treat():
    bakers_dozen = [

>         "glazed donut", "old-fashioned", "jelly donut", "bear claw", "classic donut with pink frosting and sprinkles on top", "donut hole", "another glazed donut", "yet another glazed donut?", "cruller", "maple donut", "powdered sugar donut", "doughnut", "thirteenth donut"

Reviewer: How is it that "Bismark" doesn't make the list of donuts?!?!?!?!?

Author: _That's_ the name of those?!?!?!?!?!? 😆

        thirteenth_donut = get_celebration_treat(bakers_dozen)
        self.assertEqual(thirteenth_donut, "celebration thirteenth donut!")

    def test_no_celebration_treat(self):
        empty_donut_box = []
        self.assertNotEqual(len(empty_donut_box), 13)
        no_donut = get_celebration_treat(empty_donut_box)
        self.assertEqual(no_donut, "no donut :(")

    def test_dont_grab_thirteenth_donut(self):
        # This test is for checking that grab_thirteenth_donut raises the error
        empty_donut_box = []
        with self.assertRaises(IndexError):
            grab_thirteenth_donut(empty_donut_box)
```
### !end-tests
### !explanation

An example of a working implementation:

```python
def get_celebration_treat(treats):
    try:
        thirteenth_donut = grab_thirteenth_donut(treats)
        return f"celebration {thirteenth_donut}!"
    except IndexError:
        return "no donut :("
```

### !end-explanation
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question 5 -->

<!-- prettier-ignore-start -->
### !challenge

* type: code-snippet
* language: python3.6
* id: UlAHwy
* title: Exception Handling

### !question

> Given these tests, refactor **_only_** the **`prepare_meal`** function so it will pass. Use `try ... except` syntax.

Reviewer: Similar rephrasing as above.
```suggestion
Given these tests, refactor **_only_** the **`prepare_meal`** function to make them pass. Use `try ... except` syntax.
```
